src/main/java/ee/tuleva/onboarding/mandate/email/MandateEmailSender.java

```diff
@@ -18,22 +18,20 @@
   private final UserConversionService conversionService;
 
   @EventListener
-  public void sendEmail(SecondPillarAfterMandateSignedEvent event) {
+  public void sendEmail(AfterMandateSignedEvent event) {
     UserPreferences contactDetails = episService.getContactDetails(event.getUser());
-    ConversionResponse conversion = conversionService.getConversion(event.getUser());
-    PillarSuggestion pillarSuggestion =
-        new PillarSuggestion(3, event.getUser(), contactDetails, conversion);
+    PillarSuggestion pillarSuggestion = buildPillarSuggestion(event, contactDetails);
     emailService.sendMandate(
         event.getUser(), event.getMandate(), pillarSuggestion, contactDetails, event.getLocale());
   }
 
-  @EventListener
-  public void sendEmail(ThirdPillarAfterMandateSignedEvent event) {
-    UserPreferences contactDetails = episService.getContactDetails(event.getUser());
+  private PillarSuggestion buildPillarSuggestion(
+      AfterMandateSignedEvent event, UserPreferences contactDetails) {
     ConversionResponse conversion = conversionService.getConversion(event.getUser());
-    PillarSuggestion pillarSuggestion =
-        new PillarSuggestion(2, event.getUser(), contactDetails, conversion);
-    emailService.sendMandate(
-        event.getUser(), event.getMandate(), pillarSuggestion, contactDetails, event.getLocale());
+    return new PillarSuggestion(
+        event instanceof ThirdPillarAfterMandateSignedEvent ? 2 : 3,
```

Comment:
I avoid using `instanceof` at all costs. Usually, it means that I'm breaking the **polymorphism** principle of Object Oriented Programming.

Any way to solve this polymorphically instead? i.e. `event.getPillar()`?

Reply:
I would maybe try to get rid of the `SecondPillarAfterMandateSignedEvent` and `ThirdPillarAfterMandateSignedEvent` subclasses and add a `getPillar()` method directly to `AfterMandateSignedEvent`.